osu.Game/Storyboards/StoryboardSprite.cs

```diff
@@ -209,38 +209,61 @@ private static double calculateOptimisedEndTime(CommandTimelineGroup timelineGro
         {
             // Here we are starting from maximum value and trying to minimise the end time on each step.
             // There are few solid guesses we can make using which sprite's end time can be minimised: alpha = 0, scale = 0, colour.a = 0.
+            // If one of these conditions is met, there's no point of keeping the sprite alive since it will be invisible regardless.
             double[] deathTimes =
             {
                 double.MaxValue, // alpha
                 double.MaxValue, // colour alpha
-                double.MaxValue, // scale
+                double.MaxValue, // vector scale
                 double.MaxValue, // scale x
                 double.MaxValue, // scale y
             };
 
-            // The loops below are following the same pattern.
-            // We could be using TimelineGroup.EndValue here, however it's possible to have multiple commands with 0 value in a row
-            // so we are saving the earliest of them.
+            // Loops below are following the same pattern.
             foreach (var alphaCommand in timelineGroup.Alpha.Commands)
             {
+                // sprite is invisible by the end of the command - check the start value
                 if (alphaCommand.EndValue == 0)
-                    // commands are ordered by the start time, however end time may vary. Save the earliest.
-                    deathTimes[0] = Math.Min(alphaCommand.EndTime, deathTimes[0]);
+                {
+                    if (alphaCommand.StartValue > 0)
+                    {
+                        // sprite is visible at the start and hidden at the end - should not die earlier than the end of this command
+                        deathTimes[0] = alphaCommand.EndTime;
```

Comment:
Shouldn't this be `Math.Max(deathTImes[0], alphaCommand.EndTime)` for safety or something?

Also goes for every single similar usage pattern below.

Reply:
As far as my understanding goes, no. Since commands are ordered by the start time and new command will apply start value immediately - it will be overriden regardless of the length of the previous commands and there's no reason to compare new time with the saved one.
Example:
```
F,0,0,10000,0,1 // Fade from 0 to 1 starting from 0 with 10s duration
F,0,1000,2000,1,0 // Fade from 1 to 0 starting from 1s with 1s duration
```
Even if the 1st command is 10s long and 2nd command starts while 1st is active, 2nd command will become active at 1s mark and after that there's no need to keep track of the 1st one.
With my implementation end time will be 2000, while with yours optimisation will be skipped completely `Max(double.MaxValue, 2000) = double.MaxValue`

Reply:
I'm gonna have to go cross-check stable to confirm that.

Reply:
So I cross-checked stable and what you say is pretty blatantly false, and it doesn't take access to source to prove it.

Here's a beatmap: [a - a.zip](https://github.com/ppy/osu/files/14851155/a.-.a.zip) (rename to `.osz` to import).

Inside the storyboard, I put this:

```
Sprite,Foreground,Centre,"bug.png",320,240
 F,0,1200,4000,0,1
 F,0,2500,3000,1,0
```

Stable does this:

https://github.com/ppy/osu/assets/20418176/ee087b3f-3bd7-4672-a790-236935c9d415

and lazer does this:

https://github.com/ppy/osu/assets/20418176/e1b05a27-ca7d-4db9-b87a-87e9cb1427db

The long and short of it is, no, stable does not discard the first transform if it sees there are two concurrent fade transforms. It discards _the second one_. As in, the _first transform is played out start to end_ and the second one is _ignored_.

As far as I can tell, the stable code for this looks a bit like so:

```csharp
            // First check for active transformations
            foreach (Transformation t in Transformations)
            {
                if (t.Time1 >= time || t.Time2 > time)
                {
                    hasFuture = true;

                    if (t.Loop && t.CurrentLoopCount > 0 && !SpriteManager.Current.ForwardPlayOptimisations)
                        hasUpdateableLoops = true;
                    if (t.Time1 > time)
                        continue;
                }

                if (t.Time2 <= time)
                {
                    hasPast = true;

                    if (t.Loop && (t.MaxLoopCount == 0 || t.CurrentLoopCount < t.MaxLoopCount - 1))
                        hasUpdateableLoops = true;
                    if (t.Time2 < time)
                        continue;
                }

                shouldDraw = true;

                switch (t.Type)
                {
                    case TransformationType.Fade:
                        if (hasFade) break;
                        Alpha = OsuMathHelper.TweenValues(t.StartFloat, t.EndFloat, time, t.Time1, t.Time2, t.Easing);
                        hasFade = true;
                        break;
```

(https://github.com/peppy/osu-stable-reference/blob/e53980dd76857ee899f66ce519ba1597e7874f28/osu!/Graphics/Sprites/pSprite.cs#L532-L563)

Long and short of it is, if there are multiple concurrent transforms from a single group, the _first one_ wins. Now the fact that the display is different is _two_ bugs, one of which is in the _application_ of the fades and thus out of scope here, but it still makes this variant of the optimisation _wrong_, since the lifetime end of the sprite will be calculated to be 3000 with this code, while in fact it should be no less than 4000.

I'm somewhat miffed that I am the one having to verify this experimentally. Yes I may have stable access but all it takes to check this is having access to stable and making a test case to test this.

Reply:
> So I cross-checked stable and what you say is pretty blatantly false

I never said a word about stable. Stable does handle concurent transforms differently and that's [a separate issue](https://github.com/ppy/osu/issues/7257#issuecomment-567844226). My approach aims for the lazer transforms (the way they are implemented currently). Though I'm not against fixing those first and then looking into optimizing and leaving this pr on-hold until then. Or may be better yet just revert my older changes since they are pretty isolated.

Reply:
there should be a todo at least

Reply:
Am I understanding that this was already broken in master prior to the optimistion PR?

Reply:
This pr is a fix for a regression caused by my prior changes.
The "todo" isn't related to the issue I'm trying to resolve and is here because while this code works as expected right now, it won't be entirely correct when (or if) concurrent transforms handling will be adjusted to match stable.

Reply:
But didn't the previous code already consider exactly the scenario @bdach is mentioning? It used `CommandsEndTime` which is `max(EndTime)`, and your optimised version changes that to the earliest time?